lib/std/crypto/ml_kem.zig

```diff
@@ -229,8 +229,6 @@ fn Kyber(comptime p: Params) type {
         pub const shared_length = common_shared_key_size;
         /// Length (in bytes) of a seed for deterministic encapsulation.
         pub const encaps_seed_length = common_encaps_seed_length;
-        /// Length (in bytes) of a seed for key generation.
-        pub const seed_length: usize = inner_seed_length + shared_length;
```

Comment:
Was moving`seed_length` necessary?

This is not consistent. For example, `encaps_seed_length` is only used in `PublicKey`, so to be consistent, it should be moved there as well.

Having all the parameters at the same place, at the root level of the primitive, feels less surprising.

`seed_length` is in `KeyPair` in `ecdsa` and `ed25519`, but I don't think it was intentional. In `X25519`, this is not the case.

Either way, we should use the same convention everywhere, so I'd like to understand why having this constant at the same level as other constants was inconvenient for your use case.

Reply:
I don't suggest such changes should be part of that PR. Just want your take on if we should try to keep all these constants together, or rather move them to the structures they are used in.

Reply:
> Was moving `seed_length` necessary?

No, but it's convenient to match `X25519.KeyPair.seed_length` and `Secp*r1.KeyPair.seed_length`.

```zig
/// One of these potential key pairs will be selected during the handshake.
pub const KeyPairs = struct {
    hello_rand: [hello_rand_length]u8,
    session_id: [session_id_length]u8,
    kyber768d00: Kyber768,
    secp256r1: Secp256r1,
    secp384r1: Secp384r1,
    x25519: X25519,

    const hello_rand_length = 32;
    const session_id_length = 32;
    const X25519 = tls.NamedGroupT(.x25519).KeyPair;
    const Secp256r1 = tls.NamedGroupT(.secp256r1).KeyPair;
    const Secp384r1 = tls.NamedGroupT(.secp384r1).KeyPair;
    const Kyber768 = tls.NamedGroupT(.x25519_kyber768d00).Kyber768.KeyPair;

    pub fn init() @This() {
        var random_buffer: [
            hello_rand_length +
                session_id_length +
                Kyber768.seed_length +
                Secp256r1.seed_length +
                Secp384r1.seed_length +
                X25519.seed_length
        ]u8 = undefined;
        ...
    }
};
```

> I don't suggest such changes should be part of that PR.

Why not?

> Just want your take on if we should try to keep all these constants together, or rather move them to the structures they are used in.

I think moving them to the structures they are used in makes the most sense and is generally the pattern of the stdlib.

Reply:
> Why not?

Because this PR is already ginormous, so changes unrelated to TLS would be better handled in distinct PRs.

One thing at a time , if only to help reviewers, and carefully track side effects of breaking changes.

Reply:
> Because this PR is already ginormous, so changes unrelated to TLS would be better handled in distinct PRs.

Good point. The reader/writer changes to readv/writev are not related and are the majority of files changed. I think splitting them out would be better and I'll do that to make this PR smaller.

Reply:
On further inspection there's more crypto API changes that'd make the certificate signature signing and verification more readable than simply moving this one `seed_length` member. Having to switch on RSA modulus length everywhere is a bit nasty:
```zig
                const Hash = comptime_scheme.Hash();
                const rsa = Certificate.rsa;
                const key = rsa.PublicKey.fromDer(cert.pubKey()) catch
                    return s.writeError(.bad_certificate);
                switch (key.n.bits() / 8) {
                    inline 128, 256, 512 => |modulus_len| {
                        const sig = rsa.PSSSignature.fromBytes(modulus_len, sig_bytes);
                        rsa.PSSSignature.verify(modulus_len, sig, sig_content, key, Hash) catch
                            return s.writeError(.decode_error);
                    },
                    else => {
                        return s.writeError(.bad_certificate);
                    },
                }

```

Also `rsa` living in `Certificate` feels wrong. All other crypto APIs used in certificates live outside of it.

The `Signature` APIs for ECDSA, RSA, and ED25519 are all significantly different. It'd be good to make them similar.

I'll also split it into a separate PR.

Reply:
APIs for ECDSA and Ed25519 look quite similar.

The current RSA implementation is only for TLS and not meant to be used elsewhere.

It's meant to be eventually replaced by a generic implementation designed to support other uses cases and constructions such as [partially] blind signatures. But bits of what we already have will obviously be reused, especially all the serialization stuff.

Reply:
But yeah, the current RSA code could be moved to the `Certificate/` directory.